comment/api/permissions.py

```diff
@@ -32,10 +53,7 @@ def has_permission(self, request, view):
 
 class CanChangeFlaggedCommentState(permissions.BasePermission):
     def has_permission(self, request, view):
-        return is_comment_admin(request.user) or is_comment_moderator(request.user)
-
-    def has_object_permission(self, request, view, obj):
-        return obj.is_flagged and (is_comment_admin(request.user) or is_comment_moderator(request.user))
```

Comment:
why is the object level permission removed here? the thing `obj.is_flagged` should be required, no? everything else was anyway repeated.

Reply:
This method was never called since we are using `RetrieveAPIView`.
As per REST [documenation](https://www.django-rest-framework.org/api-guide/permissions/#custom-permissions) this needs object level view in order to be called.
However, this check is already done inside the view https://github.com/Radi85/Comment/blob/develop/comment/api/views.py#L113-L117

Reply:
Quoting the relevant section here.

> Note: The instance-level has_object_permission method will only be called if the view-level has_permission checks have already passed. Also note that in order for the instance-level checks to run, the view code should explicitly call .check_object_permissions(request, obj). If you are using the generic views then this will be handled for you by default. (Function-based views will need to check object permissions explicitly, raising PermissionDenied on failure.)

the `RetrieveAPIView` is part of `generics`.

Reply:
I have tested it several time and it's indeed not called with the RetrieveAPIView, therefore I have moved the checks inside the view.

Reply:
i just saw it wasn't working, it wasn't working because `get_object` wasn't being called inside the `view` function. it was a bug on our part. also the appropriate response should be 403 in this case and not 400.

i will post a diff so that you may verify this on your end too.

Reply:
```diff
diff --git a/comment/api/views.py b/comment/api/views.py
index b0add54..0803cab 100644
--- a/comment/api/views.py
+++ b/comment/api/views.py
@@ -108,13 +108,8 @@ class CommentDetailForFlagStateChange(generics.RetrieveAPIView):
         return context
 
     def post(self, request, *args, **kwargs):
-        comment = get_object_or_404(Comment, id=kwargs.get('pk'))
+        comment = self.get_object()
         flag = Flag.objects.get_for_comment(comment)
-        if not comment.is_flagged:
-            return Response(
-                {'detail': FlagError.REJECT_UNFLAGGED_COMMENT},
-                status=status.HTTP_400_BAD_REQUEST
-            )
         state = request.data.get('state') or request.POST.get('state')
         try:
             state = flag.get_clean_state(state)

```

After applying this patch, you should run the corresponding test with the command ` python manage.py test comment.tests.test_api.test_views.APICommentDetailForFlagStateChangeTest.test_when_comment_is_not_flagged
` to verify the bug.

Reply:
Yup calling `self.get_object()` does the trick.

Reply:
i think this bug should ideally be fixed in a separate commit. if you have the time, you may otherwise i can make one.

also, for what it's worth, i don't think `RetireveAPIView` is the best choice for this view because it is modifying the object, while the former is used for mainly read-only views.